### PR TITLE
feat: initial implementation

### PR DESCRIPTION
**Pull request by aschmahmann**
This is an initial implementation of the DNS server using CoreDNS.

The main missing feature for getting this to work is integration with a distributed database to be shared between the primary and secondary nameservers. It might be possible to do this using zone files as well, but using something like dynamodb for now seems fine.
- Note: we might want to bump the dependencies in https://github.com/ipfs/go-ds-dynamodb

### CODE REVIEW

**Review by MarcoPolo — DISMISSED** [review]
Just a first pass at the server side. I'll take a look at the client side next.

File: client/acme.go [hunk 1]
```diff
@@ -291,20 +302,13 @@ func (m *P2PForgeCertMgr) Libp2pOptions(opts ...P2PForgeHostOptions) libp2p.Opti
...
-	return libp2p.ChainOptions(
-		libp2p.AddrsFactory(addrFactoryFn(func() peer.ID { return m.hostFn().ID() }, forgeDomain, cfg.allowPrivateForgeAddrs, p2pForgeWssComponent)),
-		// If we had an Fx option we could do this:
-		libp2p.WithFxOption(
-			fx.Invoke(func(host host.Host) {
-				m.provideHost(host)
-			}),
-		),
-	)
+	return libp2p.AddrsFactory(addrFactoryFn(m.hasHost, func() peer.ID { return m.hostFn().ID() }, forgeDomain, cfg.allowPrivateForgeAddrs, p2pForgeWssComponent))
```
[the remark this answers is not in the document]
Comment (MarcoPolo) [reply client/acme.go]: That makes sense.

But I want to change this to get rid of the addr factory :)

File: README.md [hunk 2]
```diff
@@ -1,3 +1,142 @@
...
+   - the [Identify protocol](https://github.com/libp2p/specs/tree/master/identify) (`/ipfs/id/1.0.0`)
+
+To set an ACME challenge send an HTTP request to the server (for libp2p.direct this is registration.libp2p.direct)
+```shell
+curl -X POST "https://registration.libp2p.direct/v1/<peerID>/_acme-challenge" \
+-H "Authorization: libp2p-PeerID bearer=\"<base64-encoded-opaque-blob>\""
+-H "Content-Type: application/json" \
+-d '{
+  "value": "your_acme_challenge_token",
+  "addresses": "[your_multiaddrs, comma_separated]"
```
Comment (achingbrain) [inline README.md:138]: Is `addresses` a string that contains an string array in JSON notation, eg:

```json
{
  "value": "...",
  "addresses": "[\"/ip4/...\", \"/ip6/...\"]"
}
```

or is it a single string, eg:

```json
{
  "value": "...",
  "addresses": "/ip4/...,/ip6/..."
}
```

If so, why not just a simple array of strings, eg:

```json
{
  "value": "...",
  "addresses": ["/ip4/...", "/ip6/..."]
}
```

Comment (achingbrain) [reply README.md:138]: Looking at [the code](https://github.com/ipshipyard/p2p-forge/pull/1/files#diff-1db9aa3c1835054d2178f88673e32e5f18ed6b3a8e787df6865b89fe3e9b90eeR51) I think it is an array?

The example might be better as:

```suggestion
  "addresses": ["your", "multiaddrs"]
```

File: README.md [hunk 3]
```diff
@@ -1,3 +1,142 @@
...
+}'
+```
+
+Where the bearer token is derived via the [libp2p HTTP PeerID Auth Specification](https://github.com/libp2p/specs/pull/564).
```
Comment (achingbrain) [inline README.md:142]: The spec can be linked to directly:


```suggestion
Where the bearer token is derived via the [libp2p HTTP PeerID Auth Specification](https://github.com/libp2p/specs/blob/master/http/README.md).
```

File: README.md [hunk 4]
```diff
@@ -1,3 +1,142 @@
...
+   - one of the following libp2p transport configurations:
+     - QUIC-v1
+     - TCP or WS or WSS, Yamux, TLS or Noise
+     - WebTransport
+     - Other transports are under consideration (e.g. HTTP), if they are of interest please file an issue
+   - the [Identify protocol](https://github.com/libp2p/specs/tree/master/identify) (`/ipfs/id/1.0.0`)
+
+To set an ACME challenge send an HTTP request to the server (for libp2p.direct this is registration.libp2p.direct)
+```shell
+curl -X POST "https://registration.libp2p.direct/v1/<peerID>/_acme-challenge" \
```
Comment (achingbrain) [inline README.md:133]: The endpoint doesn't need the PeerID in it:

```suggestion
curl -X POST "https://registration.libp2p.direct/v1/_acme-challenge" \
```

File: acme/writer.go [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+			}
+
+			const ttl = time.Hour
+			err = c.Datastore.PutWithTTL(r.Context(), datastore.NewKey(peerID.String()), []byte(typedBody.Value), ttl)
+			if err != nil {
+				w.WriteHeader(http.StatusInternalServerError)
+				_, _ = w.Write([]byte(fmt.Sprintf("error storing value: %s", err)))
+				return
+			}
+			w.WriteHeader(http.StatusOK)
```
Comment (achingbrain) [inline acme/writer.go:163]: This should be `w.WriteHeader(http.StatusNoContent)` to cause a `204` since we send no content.

**Review by lidel — APPROVED** [review]
Let's merge this, as the client shipped with Kubo 0.32.0-rc1, and we want to start producing docker images from main / staging.


Please fill separate issues / PRs for unaddressed issues.

**Pull request merged**